components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt

```diff
@@ -162,6 +164,34 @@ class GeckoEngine(
         return speculativeSession ?: GeckoEngineSession(runtime, private, defaultSettings, contextId)
     }
 
+    /**
+     * See [Engine.createSessionWithDisplayMode].
+     */
+    override fun createSessionWithDisplayMode(
+        private: Boolean,
+        manifestDisplayMode: WebAppManifest.DisplayMode,
+        contextId: String?
+    ): EngineSession {
+        ThreadUtils.assertOnUiThread()
+
+        val displayMode = when (manifestDisplayMode) {
+            WebAppManifest.DisplayMode.FULLSCREEN -> GeckoSessionSettings.DISPLAY_MODE_FULLSCREEN
+            WebAppManifest.DisplayMode.STANDALONE -> GeckoSessionSettings.DISPLAY_MODE_STANDALONE
+            WebAppManifest.DisplayMode.MINIMAL_UI -> GeckoSessionSettings.DISPLAY_MODE_MINIMAL_UI
+            WebAppManifest.DisplayMode.BROWSER -> GeckoSessionSettings.DISPLAY_MODE_BROWSER
+        }
+
+        val speculativeSession = speculativeConnectionFactory.get(private, contextId)
+        return speculativeSession ?: GeckoEngineSession(runtime, private, defaultSettings, contextId, {
```

Comment:
If there's a `speculativeSession` then we would use that here and not create a new session with the given display mode. Since the speculative session will never have the right display mode currently (and I guess we will never open one with a specific display mode), we probably just want to ignore it in case we need a session with a display mode.